src/machines/java/com/enderio/machines/common/blockentity/FluidTankBlockEntity.java

```diff
@@ -90,66 +110,144 @@ public void serverTick() {
         ItemStack inputItem = getInventory().getStackInSlot(2);
         ItemStack outputItem = getInventory().getStackInSlot(3);
         if (!inputItem.isEmpty()) {
             if (inputItem.getItem() == Items.BUCKET) {
-                if (!fluidTank.isEmpty()) {
-                    FluidStack stack = fluidTank.drain(FluidType.BUCKET_VOLUME, IFluidHandler.FluidAction.SIMULATE);
-                    if (stack.getAmount() == FluidType.BUCKET_VOLUME && (outputItem.isEmpty() || (outputItem.getItem() == stack.getFluid().getBucket()
-                        && outputItem.getCount() < outputItem.getMaxStackSize()))) {
-                        fluidTank.drain(FluidType.BUCKET_VOLUME, IFluidHandler.FluidAction.EXECUTE);
-                        inputItem.shrink(1);
-                        if (outputItem.isEmpty()) {
-                            getInventory().setStackInSlot(3, stack.getFluid().getBucket().getDefaultInstance());
-                        } else {
-                            outputItem.grow(1);
-                        }
+                Pair<Boolean, FluidStack> result = drainTankWithBucket(
+                        stack -> (outputItem.isEmpty() || (outputItem.getItem() == stack.getFluid().getBucket()
+                                && outputItem.getCount() < outputItem.getMaxStackSize())));
+                if (result.getFirst()) {
+                    inputItem.shrink(1);
+                    if (outputItem.isEmpty()) {
+                        getInventory().setStackInSlot(3,
+                                result.getSecond().getFluid().getBucket().getDefaultInstance());
+                    } else {
+                        outputItem.grow(1);
                     }
                 }
             } else {
-                Optional<IFluidHandlerItem> fluidHandlerCap = inputItem.getCapability(CapabilityFluidHandler.FLUID_HANDLER_ITEM_CAPABILITY).resolve();
-                if (fluidHandlerCap.isPresent() && outputItem.isEmpty()) {
-                    IFluidHandlerItem itemFluid = fluidHandlerCap.get();
-                    int filled = moveFluids(fluidTank, itemFluid, fluidTank.getFluidAmount());
-                    if (filled > 0) {
-                        getInventory().setStackInSlot(3, itemFluid.getContainer());
+                if (outputItem.isEmpty()) {
+                    Pair<FluidOperationResult, IFluidHandlerItem> result = drainTankWithItem(inputItem);
+                    if (result.getFirst() == FluidOperationResult.SUCCESS) {
+                        getInventory().setStackInSlot(3, result.getSecond().getContainer());
                         getInventory().setStackInSlot(2, ItemStack.EMPTY);
                     }
                 }
             }
         }
     }
 
+    public Pair<Boolean, FluidStack> drainTankWithBucket() {
+        return drainTankWithBucket(null);
+    }
+
+    public Pair<Boolean, FluidStack> drainTankWithBucket(Function<FluidStack, Boolean> shouldDrain) {
```

Comment:
Mark shouldDrain as nullable if you're allowing null

Reply:
Fixed in 06c076d25b74df9093933ecb397774aed2d55f4f